File: npbloom-web/src/__tests__/util/UndoRedoHistory.test.ts
import { describe, expect, it } from 'vitest';
import UndoRedoHistory, { applyToHistory, canRedo, canUndo, redo, undo } from '../../util/UndoRedoHistory';
import { applyAction, reverseAction, TestAction, TestState } from './testActionAndState';

describe('undo/redo history', () => {
  const initialState: TestState = {
    valueNumber: 8,
    valueString: "alright very cool!",
  };
  
  const action1: TestAction = {
    type: 'SET_STRING',
    timestamp: new Date(1618033988749),
    oldValue: "alright very cool!",
    newValue: "alright cool!",
  };
  
  const action2: TestAction = {
    type: 'INCREMENT_NUMBER',
    timestamp: new Date(1618034000000),
    difference: 11,
  };

  const initialUndoRedoHistory: UndoRedoHistory<TestAction, TestState> = {
    current: initialState,
    undoStack: [],
    redoStack: [],
  };

  const undoRedoHistoryAfterTwoActions: UndoRedoHistory<TestAction, TestState> = {
    current: { ...initialState, valueNumber: 19, valueString: 'alright cool!' },
    undoStack: [action2, action1],
    redoStack: [],
  };

  const undoRedoHistoryAfterUndoOnce: UndoRedoHistory<TestAction, TestState> = {
    current: { ...initialState, valueString: 'alright cool!' },
    undoStack: [action1],
    redoStack: [action2],
  };

  const undoRedoHistoryAfterUndoTwice: UndoRedoHistory<TestAction, TestState> = {
    current: { ...initialState },
    undoStack: [],
    redoStack: [action1, action2],
  };

  it('applies two actions and writes them to history', () => {
    const resultAfterFirst = applyToHistory(applyAction)(action1)(initialUndoRedoHistory);
    const resultAfterSecond = applyToHistory(applyAction)(action2)(resultAfterFirst);
    expect(resultAfterSecond).toStrictEqual(undoRedoHistoryAfterTwoActions);
  });

  it('allows undo when there is something left to undo', () => {
    expect(canUndo(undoRedoHistoryAfterUndoOnce)).toBe(true);
  });

  it('disallows undo when there is nothing left to undo', () => {
    expect(canUndo(undoRedoHistoryAfterUndoTwice)).toBe(false);
  });

  it('allows redo when there is something left to redo', () => {
    expect(canRedo(undoRedoHistoryAfterUndoOnce)).toBe(true);
  });

  it('disallows redo when there is nothing left to redo', () => {
    expect(canRedo(undoRedoHistoryAfterTwoActions)).toBe(false);
  });

  it('undoes one action', () => {
    expect(undo(applyAction)(reverseAction)(undoRedoHistoryAfterTwoActions))
      .toStrictEqual(undoRedoHistoryAfterUndoOnce);
  });

  it('undoes two actions', () => {
    const resultAfterFirst = undo(applyAction)(reverseAction)(undoRedoHistoryAfterTwoActions);
    const resultAfterSecond = undo(applyAction)(reverseAction)(resultAfterFirst);
    expect(resultAfterSecond).toStrictEqual(undoRedoHistoryAfterUndoTwice);
  });

  it('undoes one action then redoes it', () => {
    const resultAfterUndo = undo(applyAction)(reverseAction)(undoRedoHistoryAfterTwoActions);
    const resultAfterRedo = redo(applyAction)(resultAfterUndo);
    expect(resultAfterRedo).toStrictEqual(undoRedoHistoryAfterTwoActions);
  });

  it('undoes two actions then redoes one', () => {
    const resultAfterFirstUndo = undo(applyAction)(reverseAction)(undoRedoHistoryAfterTwoActions);
    const resultAfterSecondUndo = undo(applyAction)(reverseAction)(resultAfterFirstUndo);
    const resultAfterRedo = redo(applyAction)(resultAfterSecondUndo);
    expect(resultAfterRedo).toStrictEqual(undoRedoHistoryAfterUndoOnce);
  });
});

File: npbloom-core/src/commonTest/kotlin/history/UndoRedoHistoryTest.kt
package history

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class UndoRedoHistoryTest {
    private val initialState = TestState(numberValue = 8, stringValue = "alright very cool!")

    private val action1: TestAction = SetString(oldValue = "alright very cool!", newValue = "alright cool!")

    private val action2: TestAction = IncrementNumber(difference = 11)

    private val initialUndoRedoHistory = UndoRedoHistory(::testApplyAction, ::testReverseAction, initialState)

    private val undoRedoHistoryAfterTwoActions = initialUndoRedoHistory.copy(
        current = initialState.copy(numberValue = 19, stringValue = "alright cool!"),
        undoStack = listOf(action2, action1),
    )

    private val undoRedoHistoryAfterUndoOnce = initialUndoRedoHistory.copy(
        current = initialState.copy(stringValue = "alright cool!"),
        undoStack = listOf(action1),
        redoStack = listOf(action2),
    )

    private val undoRedoHistoryAfterUndoTwice = initialUndoRedoHistory.copy(
        redoStack = listOf(action1, action2),
    )

    @Test
    fun testApplyTwoActions() =
        assertEquals(
            undoRedoHistoryAfterTwoActions,
            initialUndoRedoHistory.applyAction(action1).applyAction(action2)
        )

    @Test
    fun testApplyTwoActionsAndUndoOne() =
        assertEquals(
            undoRedoHistoryAfterUndoOnce,
            initialUndoRedoHistory.applyAction(action1).applyAction(action2).undo()
        )

    @Test
    fun testApplyTwoActionsAndUndoBoth() =
        assertEquals(
            undoRedoHistoryAfterUndoTwice,
            initialUndoRedoHistory.applyAction(action1).applyAction(action2).undo().undo()
        )

    @Test
    fun testApplyTwoActionsUndoOneThenRedoIt() =
        assertEquals(
            undoRedoHistoryAfterTwoActions,
            initialUndoRedoHistory.applyAction(action1).applyAction(action2).undo().redo()
        )

    @Test
    fun testApplyTwoActionsUndoBothThenRedoOne() =
        assertEquals(
            undoRedoHistoryAfterUndoOnce,
            initialUndoRedoHistory.applyAction(action1).applyAction(action2).undo().undo().redo()
        )

    @Test
    fun testCanUndoTrue() =
        assertTrue(undoRedoHistoryAfterUndoOnce.canUndo)

    @Test
    fun testCanUndoFalse() =
        assertFalse(undoRedoHistoryAfterUndoTwice.canUndo)

    @Test
    fun testCanRedoTrue() =
        assertTrue(undoRedoHistoryAfterUndoOnce.canRedo)

    @Test
    fun testCanRedoFalse() =
        assertFalse(undoRedoHistoryAfterTwoActions.canRedo)
}
